### PR TITLE
Allow users to filter applications by SAML issuer

### PR DESCRIPTION
**Pull request by ravinduramesh**
### Proposed changes in this pull request
- Allow filter query to contain any number of 'and', 'or' operations.
- Added SAML issuer as an application filter attribute.

### Related Issues
Fixes [asgardeo-product/#11894](https://github.com/wso2-enterprise/asgardeo-product/issues/11894)

## Related PRs
Merge after merging [carbon-identity-framework/#4240](https://github.com/wso2/carbon-identity-framework/pull/4240)

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/ServerApplicationManagementService.java [hunk 1]
```diff
@@ -157,33 +158,26 @@ public class ServerApplicationManagementService {
...
-
     // WS-Trust related constants.
     private static final String WS_TRUST_TEMPLATE_ID = "061a3de4-8c08-4878-84a6-24245f11bf0e";
     private static final String STS_TEMPLATE_NOT_FOUND_MESSAGE = "Request template with id: %s could " +
             "not be found since the WS-Trust connector has not been configured.";
 
     static {
-        SEARCH_SUPPORTED_FIELD_MAP.put("name", "SP_APP.APP_NAME");
-        SEARCH_SUPPORTED_FIELD_MAP.put("clientId", "SP_INBOUND_AUTH.INBOUND_AUTH_KEY");
+        SUPPORTED_FILTER_ATTRIBUTES.add("name");
```
Comment (ravinduramesh) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/ServerApplicationManagementService.java:171]: Create a constant for "name" also.

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 2]
```diff
@@ -2121,8 +2121,9 @@ components:
       name: filter
       required: false
       description: |
-        Condition to filter the retrieval of records. Supports 'sw', 'co', 'ew' and 'eq' operations.
-        Currently supports only filtering based on the 'name' and the 'clientId' attribute.
+        Condition to filter the retrieval of records.
+        Supports 'sw', 'co', 'ew', and 'eq' operations with 'and', 'or' logical operations.
+        Currently supports only filtering based on the 'name', the 'clientId', and the 'issuer' attributes.
```
Comment (ravinduramesh) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:2126]: Mention the precedence of logical operators.

Comment (ravinduramesh) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:2126]: With some sample

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/ServerApplicationManagementService.java [hunk 3]
```diff
@@ -307,6 +264,36 @@ public ApplicationListResponse getAllApplications(Integer limit, Integer offset,
...
+            if (operationNode.getOperation().equals("not")) {
+                throw buildClientError(ErrorMessage.INVALID_FILTER_OPERATION);
+            }
+
+            return Stream.of(validateFilterTree(leftNode), validateFilterTree(rightNode))
+                    .flatMap(Collection::stream)
+                    .collect(Collectors.toList());
+        } else {
+            throw buildClientError(ErrorMessage.INVALID_FILTER_FORMAT);
+        }
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/ServerApplicationManagementService.java:294]:
```suggestion
        throw buildClientError(ErrorMessage.INVALID_FILTER_FORMAT);
```

**Review by kayathiri4 — APPROVED** [review]

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 4]
```diff
@@ -2121,10 +2121,14 @@ components:
...
-        Condition to filter the retrieval of records. Supports 'sw', 'co', 'ew' and 'eq' operations.
-        Currently supports only filtering based on the 'name' and the 'clientId' attribute.
+        Condition to filter the retrieval of records.
+        Supports 'sw', 'co', 'ew', and 'eq' operations with 'and', 'or' logical operators.
+        Please note that 'and' and 'or' operators in filters follow the general precedence of logical operators
+        ex: A and B or C and D = (A and B) or (C and D)).
+        Currently supports only filtering based on the 'name', the 'clientId', and the 'issuer' attributes.
 
         /applications?filter=name+eq+user_portal
+        <br>
```
Comment (Achintha444) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:2131]: can we replace <br> with <br />

**Pull request merged**